src/Microsoft.AspNet.Cors.Core/CorsEngine.cs

```diff
@@ -0,0 +1,187 @@
+        /// </summary>
+        /// <param name="options">The option model representing <see cref="CorsOptions"/>.</param>
+        public CorsEngine([NotNull] IOptions<CorsOptions> options)
+        {
+            _options = options.Options;
+        }
+
+        /// <summary>
+        /// Looks up a policy using the <paramref name="policyName"/> and then evaluates the policy using the passed in
+        /// <paramref name="context"/>.
+        /// </summary>
+        /// <param name="requestContext"></param>
+        /// <param name="policyName"></param>
+        /// <returns>A <see cref="CorsResult"/> which contains the result of policy evaluation and can be
+        /// used by the caller to set apporpriate response headers.</returns>
+        public CorsResult EvaluatePolicy([NotNull] HttpContext context, string policyName)
+        {
+            var policy = _options.GetPolicy(policyName);
+            return EvaluatePolicy(context, policy);
+        }
+
+        /// <inheritdoc />
+        public CorsResult EvaluatePolicy([NotNull] HttpContext context, [NotNull]CorsPolicy policy)
+        {
+            var corsResult = new CorsResult();
+
+            if (!TryValidateOrigin(context, policy, corsResult))
+            {
+                return corsResult;
+            }
+
+            corsResult.SupportsCredentials = policy.SupportsCredentials;
+            if (context.Request.IsPreflight())
+            {
+                if (!TryValidateMethod(context, policy, corsResult))
+                {
+                    return corsResult;
+                }
+
+                if (!TryValidateHeaders(context, policy, corsResult))
+                {
+                    return corsResult;
+                }
+
+                corsResult.PreflightMaxAge = policy.PreflightMaxAge;
+            }
+            else
+            {
+                AddHeaderValues(corsResult.AllowedExposedHeaders, policy.ExposedHeaders);
+            }
+
+            return corsResult;
+        }
+
+        /// <summary>
+        /// Try to validate the request origin based on <see cref="CorsPolicy"/>.
+        /// </summary>
+        /// <param name="context">The <see cref="HttpContext"/>.</param>
+        /// <param name="policy">The <see cref="CorsPolicy"/>.</param>
+        /// <param name="result">The <see cref="CorsResult"/>.</param>
+        /// <returns><c>true</c> if the request origin is valid; otherwise, <c>false</c>.</returns>
+        public virtual bool TryValidateOrigin(
+            [NotNull] HttpContext context,
+            [NotNull] CorsPolicy policy,
+            [NotNull] CorsResult result)
+        {
+            var origin = context.Request.Headers.Get(CorsConstants.Origin);
+            if (origin != null)
+            {
+                if (policy.AllowAnyOrigin)
+                {
+                    if (policy.SupportsCredentials)
+                    {
+                        result.AllowedOrigin = origin;
```

Comment:
Should we also include `Vary: Origin` in the result?

http://www.w3.org/TR/cors/#resource-implementation